### PR TITLE
adding go-cfclient

### PR DESCRIPTION
**Pull request by sleungcy**
@Cocossoul 
@loafoe 

- I added the dependency to go-cfclient, and an authenticated instance of it into runbinder.
- Added initial upgrade to v3 for resource_cf_route_service_binding through using the go-cfclient.

### CODE REVIEW

File: cloudfoundry/managers/session.go [hunk 1]
```diff
@@ -221,30 +225,35 @@ func (s *Session) init(config *configv3.Config, configUaa *configv3.Config, conf
...
 		errType = "SSO passcode"
+		goClientConfigOptions = goConfig.Token(accessToken, refreshToken)
 	} else if config.CFUsername() != "" {
 		// try connecting with pair given on uaa to retrieve access token and refresh token
 		accessToken, refreshToken, err = uaaClient.Authenticate(map[string]string{
 			"username": config.CFUsername(),
 			"password": config.CFPassword(),
 		}, configSess.Origin, constant.GrantTypePassword)
 		errType = "username/password"
+		goClientConfigOptions = goConfig.UserPassword(config.CFUsername(), config.CFPassword())
```
Comment (Cocossoul) [inline cloudfoundry/managers/session.go:249]: Why not use the token like above ?

Maybe it's best to get the token with the old api for now, and set the goClientConfigOptions at the end with the token like you do line 241.

**Comment by Cocossoul** [comment]
Since api v3 the route service binding ID was introduced.
It is a breaking change but I think it would be best if the resource ID is the route service binding ID, and not the combination of route id and service instance id like it is now.
It is clearer, and it allows to avoid having to fetch the id again to delete the resource

File: cloudfoundry/resource_cf_route_service_binding.go [hunk 2]
```diff
@@ -79,13 +107,18 @@ func resourceRouteServiceBindingRead(ctx context.Context, d *schema.ResourceData
 	if err != nil {
 		return diag.FromErr(err)
 	}
-	routes, _, err := session.ClientV2.GetServiceBindingRoutes(serviceID)
+
+	options := client.NewServiceRouteBindingListOptions()
+	options.ServiceInstanceGUIDs = client.Filter{Values: []string{serviceID}}
+
+	routeBindings, err := session.ClientGo.ServiceRouteBindings.ListAll(context.Background(), options)
```
Comment (Cocossoul) [inline cloudfoundry/resource_cf_route_service_binding.go:114]:
```suggestion

	options := client.NewServiceRouteBindingListOptions()
	options.ServiceInstanceGUIDs = client.Filter{Values: []string{serviceID}}
	options.RouteGUIDs = client.Filter{Values: []string{routeID}}

	routeBindings, err := session.ClientGo.ServiceRouteBindings.Single(context.Background(), options)
```

To remove the for loop below

File: cloudfoundry/resource_cf_route_service_binding.go [hunk 3]
```diff
@@ -105,6 +138,34 @@ func resourceRouteServiceBindingDelete(ctx context.Context, d *schema.ResourceDa
...
 	serviceID := d.Get("service_instance").(string)
 	routeID := d.Get("route").(string)
-	_, err := session.ClientV2.DeleteServiceBindingRoute(serviceID, routeID)
+
+	var err error
+
+	options := client.NewServiceRouteBindingListOptions()
+	options.ServiceInstanceGUIDs = client.Filter{Values: []string{serviceID}}
+	options.RouteGUIDs = client.Filter{Values: []string{routeID}}
+	routeBindings, err := session.ClientGo.ServiceRouteBindings.ListAll(context.Background(), options)
```
Comment (Cocossoul) [inline cloudfoundry/resource_cf_route_service_binding.go:147]:
```suggestion
	routeBindings, err := session.ClientGo.ServiceRouteBindings.Single(context.Background(), options)
```

File: cloudfoundry/resource_cf_route_service_binding.go [hunk 4]
```diff
@@ -105,6 +138,34 @@ func resourceRouteServiceBindingDelete(ctx context.Context, d *schema.ResourceDa
...
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	routeBindingID := ""
```
Comment (Cocossoul) [inline cloudfoundry/resource_cf_route_service_binding.go:152]: I'm not sure if it's possible to have multiple route bindings for the same route id and service instance id combo

File: cloudfoundry/managers/v3appdeployers/runbinder.go [hunk 5]
```diff
@@ -1,26 +1,31 @@
 package v3appdeployers
```
Comment (Cocossoul) [inline cloudfoundry/managers/v3appdeployers/runbinder.go:1]: I would be more confident to have two different pull requests for runbinder modification and route services, and preferably route service first since runbinder works in v3 already

**Comment by Cocossoul** [comment]
I added a lot of comments, some I could have edited directly in your code but I did not find a way to edit directly the pull request without forking your fork and creating a PR on your fork.
Do you know a better way ?

**Comment by sleungcy** [comment]
@Cocossoul let me take out the routeservice stuff from this PR

File: cloudfoundry/managers/session.go (hunk 1 above)
Comment (sleungcy) [reply cloudfoundry/managers/session.go:249]: indeed.

**Comment by sleungcy** [comment]
> Since api v3 the route service binding ID was introduced. It is a breaking change but I think it would be best if the resource ID is the route service binding ID, and not the combination of route id and service instance id like it is now. It is clearer, and it allows to avoid having to fetch the id again to delete the resource, or to read it (we can just use [GET](https://github.com/cloudfoundry/go-cfclient/blob/main/client/service_route_binding.go#L64))

Indeed. This will require implementing a stateupgrader, I'll have a look if you haven't got to it already

File: cloudfoundry/resource_cf_route_service_binding.go (hunk 2 above)
Comment (sleungcy) [reply cloudfoundry/resource_cf_route_service_binding.go:114]: I suppose once we switch to use the new V3 id, this entire code will have to change.

File: cloudfoundry/resource_cf_route_service_binding.go (hunk 4 above)
Comment (sleungcy) [reply cloudfoundry/resource_cf_route_service_binding.go:152]: I get this error when I tried it:
{"errors":[{"detail":"The route and service instance are already bound.","title":"CF-ServiceInstanceAlreadyBoundToSameRoute","code":130008}]}

**Comment by sleungcy** [comment]
will be closing and recreating this

**Comment by sleungcy** [comment]
> I added a lot of comments, some I could have edited directly in your code but I did not find a way to edit directly the pull request without forking your fork and creating a PR on your fork. Do you know a better way ?

Adding you as collaborator to my fork
